contracts/ronin/BridgeTracking.sol

```diff
@@ -112,38 +143,93 @@ contract BridgeTracking is HasBridgeContract, HasValidatorContract, Initializabl
-    }
+    _increaseBallot(_kind, _requestId, _operator, _period);
   }
 
   /**
    * Increases the ballot for the operator at a period.
    */
-  function increaseBallot(
+  function _increaseBallot(
     VoteKind _kind,
     uint256 _requestId,
     address _operator,
     uint256 _period
   ) internal {
-    if (_receiptVoted[_kind][_requestId][_operator]) {
+    ReceiptStats storage _receiptInfo = _receiptStats[_kind][_requestId];
+    if (_receiptInfo.voted[_operator]) {
       return;
     }
 
-    _totalBallots[_period]++;
-    _totalBallotsOf[_period][_operator]++;
-    _receiptVoted[_kind][_requestId][_operator] = true;
+    _receiptInfo.voted[_operator] = true;
+
+    // Only records within a period
+    if (_receiptInfo.approvedPeriod == _period) {
+      if (_temporaryStats.info.totalBallotsOf[_operator] == 0) {
+        _temporaryStats.info.voters.push(_operator);
+      }
+      _temporaryStats.info.totalBallots++;
+      _temporaryStats.info.totalBallotsOf[_operator]++;
+    }
+  }
+
+  /**
+   * @dev See `totalBallotsOf`.
+   */
+  function _totalBallotsOf(
+    uint256 _period,
+    address _bridgeOperator,
+    bool _mustCountLastStats
+  ) internal view returns (uint256 _totalBallots) {
+    _totalBallots = _periodStats[_period].totalBallotsOf[_bridgeOperator];
+    if (_mustCountLastStats) {
+      _totalBallots += _temporaryStats.info.totalBallotsOf[_bridgeOperator];
+    }
+  }
+
+  /**
+   * @dev Syncs period stats if the last epoch + 1 is already wrapped up.
+   */
+  function _trySyncPeriodStats() internal {
+    uint256 _currentEpoch = _validatorContract.epochOf(block.number);
+    if (_temporaryStats.lastEpoch < _currentEpoch) {
+      (bool _filled, uint256 _period) = _validatorContract.tryGetPeriodOfEpoch(_temporaryStats.lastEpoch + 1);
+      if (!_filled) {
+        return;
+      }
+
+      VoteStats storage _stats = _periodStats[_period];
+      _stats.totalVotes += _temporaryStats.info.totalVotes;
+      _stats.totalBallots += _temporaryStats.info.totalBallots;
+
+      address _voter;
+      for (uint _i = 0; _i < _temporaryStats.info.voters.length; _i++) {
+        _voter = _temporaryStats.info.voters[_i];
+        _stats.totalBallotsOf[_voter] += _temporaryStats.info.totalBallotsOf[_voter];
+        delete _temporaryStats.info.totalBallotsOf[_voter];
+      }
+      delete _temporaryStats.info;
+      _temporaryStats.lastEpoch = _currentEpoch;
+    }
+  }
+
+  /**
+   * @dev Returns whether the last stats must be counted or not;
+   */
+  function _isLastStatsCountedForPeriod(uint256 _queriedPeriod) internal view returns (bool) {
+    uint256 _currentEpoch = _validatorContract.epochOf(block.number);
+    (bool _filled, uint256 _periodOfNextTemporaryEpoch) = _validatorContract.tryGetPeriodOfEpoch(
+      _temporaryStats.lastEpoch + 1
+    );
+    return _filled && _queriedPeriod == _periodOfNextTemporaryEpoch && _temporaryStats.lastEpoch < _currentEpoch;
```

Comment:
IMO, I think we can determine whether to count the last updated stats (or so-called buffer) in the `wrapUpEpoch` function. The validator contract should totally determine the time control, and the number of the period (or the "next" period) should not be predicted. Implementing the counting by buffering mechanism could be a better approach.